// tldraw/dist-esm/lib/ui/components/Toolbar/DefaultToolbarContent.mjs
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { GeoShapeGeoStyle, useEditor, useValue } from "@tldraw/editor";
import { useTools } from "../../hooks/useTools.mjs";
import { TldrawUiMenuItem } from "../primitives/menus/TldrawUiMenuItem.mjs";
function DefaultToolbarContent() {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(SelectToolbarItem, {}),
    /* @__PURE__ */ jsx(HandToolbarItem, {}),
    /* @__PURE__ */ jsx(DrawToolbarItem, {}),
    /* @__PURE__ */ jsx(EraserToolbarItem, {}),
    /* @__PURE__ */ jsx(ArrowToolbarItem, {}),
    /* @__PURE__ */ jsx(TextToolbarItem, {}),
    /* @__PURE__ */ jsx(NoteToolbarItem, {}),
    /* @__PURE__ */ jsx(AssetToolbarItem, {}),
    /* @__PURE__ */ jsx(RectangleToolbarItem, {}),
    /* @__PURE__ */ jsx(EllipseToolbarItem, {}),
    /* @__PURE__ */ jsx(DiamondToolbarItem, {}),
    /* @__PURE__ */ jsx(TriangleToolbarItem, {}),
    /* @__PURE__ */ jsx(TrapezoidToolbarItem, {}),
    /* @__PURE__ */ jsx(RhombusToolbarItem, {}),
    /* @__PURE__ */ jsx(HexagonToolbarItem, {}),
    /* @__PURE__ */ jsx(CloudToolbarItem, {}),
    /* @__PURE__ */ jsx(StarToolbarItem, {}),
    /* @__PURE__ */ jsx(OvalToolbarItem, {}),
    /* @__PURE__ */ jsx(XBoxToolbarItem, {}),
    /* @__PURE__ */ jsx(CheckBoxToolbarItem, {}),
    /* @__PURE__ */ jsx(ArrowLeftToolbarItem, {}),
    /* @__PURE__ */ jsx(ArrowUpToolbarItem, {}),
    /* @__PURE__ */ jsx(ArrowDownToolbarItem, {}),
    /* @__PURE__ */ jsx(ArrowRightToolbarItem, {}),
    /* @__PURE__ */ jsx(LineToolbarItem, {}),
    /* @__PURE__ */ jsx(HighlightToolbarItem, {}),
    /* @__PURE__ */ jsx(FrameToolbarItem, {}),
    /* @__PURE__ */ jsx(LaserToolbarItem, {})
  ] });
}
function useIsToolSelected(tool) {
  const editor = useEditor();
  const geo = tool.meta?.geo;
  return useValue(
    "is tool selected",
    () => {
      const activeToolId = editor.getCurrentToolId();
      const geoState = editor.getSharedStyles().getAsKnownValue(GeoShapeGeoStyle);
      return geo ? activeToolId === "geo" && geoState === geo : activeToolId === tool.id;
    },
    [editor, tool.id, geo]
  );
}
function SelectToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["select"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["select"], isSelected });
}
function HandToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["hand"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["hand"], isSelected });
}
function DrawToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["draw"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["draw"], isSelected });
}
function EraserToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["eraser"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["eraser"], isSelected });
}
function ArrowToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["arrow"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["arrow"], isSelected });
}
function TextToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["text"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["text"], isSelected });
}
function NoteToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["note"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["note"], isSelected });
}
function AssetToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["asset"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["asset"], isSelected });
}
function RectangleToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["rectangle"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["rectangle"], isSelected });
}
function EllipseToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["ellipse"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["ellipse"], isSelected });
}
function DiamondToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["diamond"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["diamond"], isSelected });
}
function TriangleToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["triangle"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["triangle"], isSelected });
}
function TrapezoidToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["trapezoid"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["trapezoid"], isSelected });
}
function RhombusToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["rhombus"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["rhombus"], isSelected });
}
function HexagonToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["hexagon"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["hexagon"], isSelected });
}
function CloudToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["cloud"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["cloud"], isSelected });
}
function StarToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["star"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["star"], isSelected });
}
function OvalToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["oval"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["oval"], isSelected });
}
function XBoxToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["x-box"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["x-box"], isSelected });
}
function CheckBoxToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["check-box"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["check-box"], isSelected });
}
function ArrowLeftToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["arrow-left"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["arrow-left"], isSelected });
}
function ArrowUpToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["arrow-up"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["arrow-up"], isSelected });
}
function ArrowDownToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["arrow-down"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["arrow-down"], isSelected });
}
function ArrowRightToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["arrow-right"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["arrow-right"], isSelected });
}
function LineToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["line"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["line"], isSelected });
}
function HighlightToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["highlight"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["highlight"], isSelected });
}
function FrameToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["frame"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["frame"], isSelected });
}
function LaserToolbarItem() {
  const tools = useTools();
  const isSelected = useIsToolSelected(tools["laser"]);
  return /* @__PURE__ */ jsx(TldrawUiMenuItem, { ...tools["laser"], isSelected });
}
export {
  ArrowDownToolbarItem,
  ArrowLeftToolbarItem,
  ArrowRightToolbarItem,
  ArrowToolbarItem,
  ArrowUpToolbarItem,
  AssetToolbarItem,
  CheckBoxToolbarItem,
  CloudToolbarItem,
  DefaultToolbarContent,
  DiamondToolbarItem,
  DrawToolbarItem,
  EllipseToolbarItem,
  EraserToolbarItem,
  FrameToolbarItem,
  HandToolbarItem,
  HexagonToolbarItem,
  HighlightToolbarItem,
  LaserToolbarItem,
  LineToolbarItem,
  NoteToolbarItem,
  OvalToolbarItem,
  RectangleToolbarItem,
  RhombusToolbarItem,
  SelectToolbarItem,
  StarToolbarItem,
  TextToolbarItem,
  TrapezoidToolbarItem,
  TriangleToolbarItem,
  XBoxToolbarItem,
  useIsToolSelected
};
