// tldraw/dist-cjs/lib/ui/components/Toolbar/DefaultToolbarContent.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DefaultToolbarContent_exports = {};
__export(DefaultToolbarContent_exports, {
  ArrowDownToolbarItem: () => ArrowDownToolbarItem,
  ArrowLeftToolbarItem: () => ArrowLeftToolbarItem,
  ArrowRightToolbarItem: () => ArrowRightToolbarItem,
  ArrowToolbarItem: () => ArrowToolbarItem,
  ArrowUpToolbarItem: () => ArrowUpToolbarItem,
  AssetToolbarItem: () => AssetToolbarItem,
  CheckBoxToolbarItem: () => CheckBoxToolbarItem,
  CloudToolbarItem: () => CloudToolbarItem,
  DefaultToolbarContent: () => DefaultToolbarContent,
  DiamondToolbarItem: () => DiamondToolbarItem,
  DrawToolbarItem: () => DrawToolbarItem,
  EllipseToolbarItem: () => EllipseToolbarItem,
  EraserToolbarItem: () => EraserToolbarItem,
  FrameToolbarItem: () => FrameToolbarItem,
  HandToolbarItem: () => HandToolbarItem,
  HexagonToolbarItem: () => HexagonToolbarItem,
  HighlightToolbarItem: () => HighlightToolbarItem,
  LaserToolbarItem: () => LaserToolbarItem,
  LineToolbarItem: () => LineToolbarItem,
  NoteToolbarItem: () => NoteToolbarItem,
  OvalToolbarItem: () => OvalToolbarItem,
  RectangleToolbarItem: () => RectangleToolbarItem,
  RhombusToolbarItem: () => RhombusToolbarItem,
  SelectToolbarItem: () => SelectToolbarItem,
  StarToolbarItem: () => StarToolbarItem,
  TextToolbarItem: () => TextToolbarItem,
  TrapezoidToolbarItem: () => TrapezoidToolbarItem,
  TriangleToolbarItem: () => TriangleToolbarItem,
  XBoxToolbarItem: () => XBoxToolbarItem,
  useIsToolSelected: () => useIsToolSelected
});
module.exports = __toCommonJS(DefaultToolbarContent_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_useTools = require("../../hooks/useTools");
var import_TldrawUiMenuItem = require("../primitives/menus/TldrawUiMenuItem");
function DefaultToolbarContent() {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SelectToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HandToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DrawToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(EraserToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ArrowToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TextToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(NoteToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AssetToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RectangleToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(EllipseToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DiamondToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TriangleToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TrapezoidToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RhombusToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HexagonToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CloudToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(StarToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(OvalToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(XBoxToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CheckBoxToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ArrowLeftToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ArrowUpToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ArrowDownToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ArrowRightToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LineToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(HighlightToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(FrameToolbarItem, {}),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LaserToolbarItem, {})
  ] });
}
function useIsToolSelected(tool) {
  const editor = (0, import_editor.useEditor)();
  const geo = tool.meta?.geo;
  return (0, import_editor.useValue)(
    "is tool selected",
    () => {
      const activeToolId = editor.getCurrentToolId();
      const geoState = editor.getSharedStyles().getAsKnownValue(import_editor.GeoShapeGeoStyle);
      return geo ? activeToolId === "geo" && geoState === geo : activeToolId === tool.id;
    },
    [editor, tool.id, geo]
  );
}
function SelectToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["select"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["select"], isSelected });
}
function HandToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["hand"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["hand"], isSelected });
}
function DrawToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["draw"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["draw"], isSelected });
}
function EraserToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["eraser"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["eraser"], isSelected });
}
function ArrowToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["arrow"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["arrow"], isSelected });
}
function TextToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["text"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["text"], isSelected });
}
function NoteToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["note"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["note"], isSelected });
}
function AssetToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["asset"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["asset"], isSelected });
}
function RectangleToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["rectangle"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["rectangle"], isSelected });
}
function EllipseToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["ellipse"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["ellipse"], isSelected });
}
function DiamondToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["diamond"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["diamond"], isSelected });
}
function TriangleToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["triangle"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["triangle"], isSelected });
}
function TrapezoidToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["trapezoid"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["trapezoid"], isSelected });
}
function RhombusToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["rhombus"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["rhombus"], isSelected });
}
function HexagonToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["hexagon"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["hexagon"], isSelected });
}
function CloudToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["cloud"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["cloud"], isSelected });
}
function StarToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["star"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["star"], isSelected });
}
function OvalToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["oval"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["oval"], isSelected });
}
function XBoxToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["x-box"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["x-box"], isSelected });
}
function CheckBoxToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["check-box"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["check-box"], isSelected });
}
function ArrowLeftToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["arrow-left"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["arrow-left"], isSelected });
}
function ArrowUpToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["arrow-up"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["arrow-up"], isSelected });
}
function ArrowDownToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["arrow-down"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["arrow-down"], isSelected });
}
function ArrowRightToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["arrow-right"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["arrow-right"], isSelected });
}
function LineToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["line"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["line"], isSelected });
}
function HighlightToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["highlight"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["highlight"], isSelected });
}
function FrameToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["frame"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["frame"], isSelected });
}
function LaserToolbarItem() {
  const tools = (0, import_useTools.useTools)();
  const isSelected = useIsToolSelected(tools["laser"]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuItem.TldrawUiMenuItem, { ...tools["laser"], isSelected });
}
