// tldraw/dist-esm/lib/shapes/note/toolStates/Pointing.mjs
import {
  StateNode,
  Vec,
  createShapeId
} from "@tldraw/editor";
import { NOTE_PIT_RADIUS, getAvailableNoteAdjacentPositions } from "../noteHelpers.mjs";
class Pointing extends StateNode {
  static id = "pointing";
  dragged = false;
  info = {};
  wasFocusedOnEnter = false;
  markId = "";
  shape = {};
  onEnter = () => {
    const { editor } = this;
    this.wasFocusedOnEnter = !editor.getIsMenuOpen();
    if (this.wasFocusedOnEnter) {
      const id = createShapeId();
      this.markId = `creating:${id}`;
      editor.mark(this.markId);
      const center = this.editor.inputs.originPagePoint.clone();
      const offset = getNotePitOffset(this.editor, center);
      if (offset) {
        center.sub(offset);
      }
      this.shape = createSticky(this.editor, id, center);
    }
  };
  onPointerMove = (info) => {
    if (this.editor.inputs.isDragging) {
      if (!this.wasFocusedOnEnter) {
        const id = createShapeId();
        const center = this.editor.inputs.originPagePoint.clone();
        const offset = getNotePitOffset(this.editor, center);
        if (offset) {
          center.sub(offset);
        }
        this.shape = createSticky(this.editor, id, center);
      }
      this.editor.setCurrentTool("select.translating", {
        ...info,
        target: "shape",
        shape: this.shape,
        onInteractionEnd: "note",
        isCreating: true,
        onCreate: () => {
          this.editor.setEditingShape(this.shape.id);
          this.editor.setCurrentTool("select.editing_shape");
        }
      });
    }
  };
  onPointerUp = () => {
    this.complete();
  };
  onInterrupt = () => {
    this.cancel();
  };
  onComplete = () => {
    this.complete();
  };
  onCancel = () => {
    this.cancel();
  };
  complete() {
    if (this.wasFocusedOnEnter) {
      if (this.editor.getInstanceState().isToolLocked) {
        this.parent.transition("idle");
      } else {
        this.editor.setEditingShape(this.shape.id);
        this.editor.setCurrentTool("select.editing_shape", {
          ...this.info,
          target: "shape",
          shape: this.shape
        });
      }
    }
  }
  cancel() {
    this.editor.bailToMark(this.markId);
    this.parent.transition("idle", this.info);
  }
}
function getNotePitOffset(editor, center) {
  let min = NOTE_PIT_RADIUS / editor.getZoomLevel();
  let offset;
  for (const pit of getAvailableNoteAdjacentPositions(editor, 0, 0)) {
    const deltaToPit = Vec.Sub(center, pit);
    const dist = deltaToPit.len();
    if (dist < min) {
      min = dist;
      offset = deltaToPit;
    }
  }
  return offset;
}
function createSticky(editor, id, center) {
  editor.createShape({
    id,
    type: "note",
    x: center.x,
    y: center.y
  }).select(id);
  const shape = editor.getShape(id);
  const bounds = editor.getShapeGeometry(shape).bounds;
  editor.updateShapes([
    {
      id,
      type: "note",
      x: shape.x - bounds.width / 2,
      y: shape.y - bounds.height / 2
    }
  ]);
  return editor.getShape(id);
}
export {
  Pointing,
  createSticky,
  getNotePitOffset
};
