// @tldraw/tlschema/dist-esm/shapes/TLTextShape.mjs
import { T } from "@tldraw/validate";
import {
  RETIRED_DOWN_MIGRATION,
  createShapePropsMigrationIds,
  createShapePropsMigrationSequence
} from "../records/TLShape.mjs";
import { DefaultColorStyle } from "../styles/TLColorStyle.mjs";
import { DefaultFontStyle } from "../styles/TLFontStyle.mjs";
import { DefaultHorizontalAlignStyle } from "../styles/TLHorizontalAlignStyle.mjs";
import { DefaultSizeStyle } from "../styles/TLSizeStyle.mjs";
const textShapeProps = {
  color: DefaultColorStyle,
  size: DefaultSizeStyle,
  font: DefaultFontStyle,
  align: DefaultHorizontalAlignStyle,
  w: T.nonZeroNumber,
  text: T.string,
  scale: T.nonZeroNumber,
  autoSize: T.boolean
};
const Versions = createShapePropsMigrationIds("text", {
  RemoveJustify: 1
});
const textShapeMigrations = createShapePropsMigrationSequence({
  sequence: [
    {
      id: Versions.RemoveJustify,
      up: (props) => {
        if (props.align === "justify") {
          props.align = "start";
        }
      },
      down: RETIRED_DOWN_MIGRATION
    }
  ]
});
export {
  textShapeMigrations,
  textShapeProps,
  Versions as textShapeVersions
};
